### PR TITLE
fix: Compute joint null mask before calling rolling corr/cov stats

### LINKED ISSUE #18217: rolling_corr giving inconsistent results
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
df1 = pl.DataFrame({'a': [1.06, 1.07, 0.93, 0.78, 0.85], 'lag_a': [1., 1.06, 1.07, 0.93, 0.78]})
df2 = pl.DataFrame({'a': [1., 1.06, 1.07, 0.93, 0.78, 0.85], 'lag_a': [None, 1., 1.06, 1.07, 0.93, 0.78]})

print(df1.select(pl.rolling_corr('a', 'lag_a', window_size=10, min_periods=5, ddof=1)))
print(df2.select(pl.rolling_corr('a', 'lag_a', window_size=10, min_periods=5, ddof=1)))
```

### Log output

```
# df1.select(pl.rolling_corr('a', 'lag_a', window_size=10, min_periods=5, ddof=1))

shape: (5, 1)
┌──────────┐
│ a        │
│ ---      │
│ f64      │
╞══════════╡
│ null     │
│ null     │
│ null     │
│ null     │
│ 0.622047 │
└──────────┘

# df2.select(pl.rolling_corr('a', 'lag_a', window_size=10, min_periods=5, ddof=1))
shape: (6, 1)
┌───────────┐
│ a         │
│ ---       │
│ f64       │
╞═══════════╡
│ null      │
│ null      │
│ null      │
│ null      │
│ null      │
│ -0.219851 │
└───────────┘
```

### Issue description

You can see that the only difference is that in `df2`, the first row for lag_a is None, because it's shifted from a.
When I compute the `rolling_corr` for both dataframes, however, I got different results.

### Expected behavior

The result from `df1`, i.e. 0.622047 is what I got from `numpy.corrcoef` as well. I wonder where the -0.219851 is coming from.

### Installed versions

<details>

```
--------Version info---------
Polars:               1.5.0
Index type:           UInt32
Platform:             macOS-14.6.1-arm64-arm-64bit
Python:               3.11.9 (main, Apr  2 2024, 08:25:04) [Clang 15.0.0 (clang-1500.3.9.4)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
great_tables:         <not installed>
hvplot:               0.10.0
matplotlib:           3.8.2
nest_asyncio:         1.6.0
numpy:                1.26.2
openpyxl:             <not installed>
pandas:               2.1.4
pyarrow:              15.0.0
pydantic:             <not installed>
pyiceberg:            <not installed>
sqlalchemy:           <not installed>
torch:                <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by agossard**
Fixes #18217

### CODE REVIEW

**Comment by agossard** [comment]
Doubt the hypothesis test failure is due to this PR. Maybe it's being addressed by #18245 @MarcoGorelli ?

File: py-polars/tests/unit/operations/rolling/test_rolling.py [hunk 1]
```diff
@@ -589,6 +589,36 @@ def test_rolling_cov_corr() -> None:
...
+    )
+
+    val_1 = df1.select(
+        pl.rolling_corr("a", "lag_a", window_size=10, min_periods=5, ddof=1).tail(1)
+    ).item()
+    val_2 = df2.select(
+        pl.rolling_corr("a", "lag_a", window_size=10, min_periods=5, ddof=1).tail(1)
+    ).item()
+
+    assert val_1 == val_2
```
Comment (ritchie46) [inline py-polars/tests/unit/operations/rolling/test_rolling.py:610]: Can you also test the actual value here?

File: py-polars/tests/unit/operations/rolling/test_rolling.py [hunk 2]
```diff
@@ -589,6 +589,36 @@ def test_rolling_cov_corr() -> None:
...
+
+    val_1 = df1.select(
+        pl.rolling_cov("a", "lag_a", window_size=10, min_periods=5, ddof=1).tail(1)
+    ).item()
+    val_2 = df2.select(
+        pl.rolling_cov("a", "lag_a", window_size=10, min_periods=5, ddof=1).tail(1)
+    ).item()
+
+    assert val_1 == val_2
```
Comment (ritchie46) [inline py-polars/tests/unit/operations/rolling/test_rolling.py:619]: Can you also test the actual value here?

**Review by ritchie46 — COMMENTED** [review]
Not entirely happy about the implementation as it leaves a lot to CSE, but it is fine for now. Let's first make it correct. 

Thanks! Got one comment about the test.

File: py-polars/tests/unit/operations/rolling/test_rolling.py (hunk 2 above)
Comment (agossard) [reply py-polars/tests/unit/operations/rolling/test_rolling.py]: I changed the test as suggested and pushed. However, I also spent some time trying to put together a hypothesis test that would cross check these corr and cov functions against numpy. I could not get it to pass, and have an example frame which yields correlation > 1.0 :-/

df = pl.DataFrame(
    {
        "a": [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0],
        "b": [101.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.000061, 0.0],
    }
)

df_corr = df.select(
    pl.rolling_corr("a", "b", window_size=7, min_periods=5, ddof=1)
)

I don't have time to push more on this right now (or even this week maybe). But I will log a separate issue.

**Pull request merged**